### PR TITLE
Npc's find near enemies

### PR DESCRIPTION
**Pull request by DavidPgl**

### CODE REVIEW

File: Assets/Scripts/Enemy/EnemyController.cs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        private float rotationTolerance;
+        
+        [Tooltip("The radius around the NPC in which a player can escape.")] [SerializeField]
+        private float escapeRadius;
+
+        [Tooltip("Show escape radius in scene view.")] [SerializeField]
+        private bool showEscapeRadius;
+
+        [Tooltip("TargetFinderComponent - Finds nearby targets")] [SerializeField]
+        private TargetFinder targetFinder;
```
Comment (DavidPgl) [inline Assets/Scripts/Enemy/EnemyController.cs]: New

File: Assets/Scripts/Enemy/EnemyController.cs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        {
+            if (!_isChasing)
+                Wander();
+
+            if (!isAggressive || targetFinder.Targets.Count == 0) return;
+
+            // Attack first target in list
+            GameObject currentTarget = targetFinder.Targets[0];
+
+            if (!_isChasing)
```
Comment (SamuelllG) [inline Assets/Scripts/Enemy/EnemyController.cs]:
```
if (!isAggressive || targetFinder.Targets.Count == 0) Wander();

// Attack first target in list
GameObject currentTarget = targetFinder.Targets[0];

if (!_isChasing) StartChasing();
```

File: Assets/Scripts/Enemy/TargetFinder.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    [RequireComponent(typeof(Collider))]
+    public class TargetFinder : MonoBehaviour
+    {
+        [Tooltip("Every layer that will be considered as possible target")] [SerializeField]
+        private LayerMask attackableLayers;
+
+        /// <summary>
+        /// List of all targets that entered the trigger-radius and have a layer of the <see cref="attackableLayers"/> layer mask
+        /// </summary>
+        public readonly List<GameObject> Targets = new List<GameObject>();  //TODO: Should probably be replaced by the abstract class "Damageable"
```
Comment (SamuelllG) [inline Assets/Scripts/Enemy/TargetFinder.cs]: Concerning the TODO: Like I said, AttackLogic needs the target's position, so passing Damageable won't be enough.

File: Assets/Scripts/Enemy/TargetFinder.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+        private void OnTriggerEnter(Collider other)
```
Comment (SamuelllG) [inline Assets/Scripts/Enemy/TargetFinder.cs]: Where do you define the radius of the trigger? It should be lookRadius.

File: Assets/Scripts/Utils/ClassExtender.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+﻿using UnityEngine;
+
+namespace Utils
+{
+    public static class ClassExtender
```
Comment (SamuelllG) [inline Assets/Scripts/Utils/ClassExtender.cs:5]: Where is this even used?

**Review by SamuelllG — CHANGES_REQUESTED** [review]

**Comment by SamuelllG** [comment]
When checking out your branch, I get several NullRefs and can't move! Additional AttackLogic scripts in prefabs all are resetted

File: Assets/Scripts/Enemy/TargetFinder.cs (hunk 4 above)
Comment (SamuelllG) [reply Assets/Scripts/Enemy/TargetFinder.cs]: Oh okay you have an additional gameObject with a trigger collider. It's a piti that you can't change the attackRadius from the enemy object anymore. At least you see the targetFinder SerializedField there.

Comment (DavidPgl) [reply Assets/Scripts/Enemy/TargetFinder.cs]: Physics.Raycast and colliders made me some porblems. That was the only way I could think of atm

File: Assets/Scripts/Utils/ClassExtender.cs (hunk 5 above)
Comment (DavidPgl) [reply Assets/Scripts/Utils/ClassExtender.cs:5]: ` if(!attackableLayers.Contains(other.gameObject.layer)) return;`
See targetFinder line 25

File: Assets/Scripts/Enemy/TargetFinder.cs (hunk 4 above)
Comment (SamuelllG) [reply Assets/Scripts/Enemy/TargetFinder.cs]: Yep, I think it's okay 👍

File: Assets/Scripts/Enemy/TargetFinder.cs (hunk 3 above)
Comment (DavidPgl) [reply Assets/Scripts/Enemy/TargetFinder.cs]: The abstract class should extend from MonoBehaviour or be an component for itself maybe?

File: Assets/Scripts/Enemy/EnemyController.cs (hunk 2 above)
Comment (DavidPgl) [reply Assets/Scripts/Enemy/EnemyController.cs]: This would prevent an aggressive enemy from wandering, right?

**Comment by DavidPgl** [comment]
^This should fix the nullrefs when going back to the main menu and starting the game again

File: Assets/Scripts/Enemy/EnemyController.cs (hunk 2 above)
Comment (Kaes3kuch3n) [reply Assets/Scripts/Enemy/EnemyController.cs]: I think this has to be reworked later to support passive, neutral and aggressive NPCs

File: Assets/Scripts/Enemy/TargetFinder.cs (hunk 3 above)
Comment (Kaes3kuch3n) [reply Assets/Scripts/Enemy/TargetFinder.cs]: How about adding something like GetPosition() to Damageable?

**Review by Kaes3kuch3n — DISMISSED** [review]

**Review by Kaes3kuch3n — APPROVED** [review]

**Review by inkompotato — APPROVED** [review]
cool

**Review by SamuelllG — APPROVED** [review]

**Pull request merged**